src/analyze.ts

```diff
@@ -399,17 +409,17 @@ module.exports = async function (id, code, job) {
   handleWrappers(ast);
   ({ ast = ast, scope = scope } = handleSpecialCases({ id, ast, scope, emitAsset: path => assets.add(path), emitAssetDirectory, job }) || {});
 
-  function backtrack (self, parent) {
+  function backtrack (self: WalkerContext, parent: Node) {
     // computing a static expression outward
     // -> compute and backtrack
     if (!staticChildNode) throw new Error('Internal error: No staticChildNode for backtrack.');
     const curStaticValue = computePureStaticValue(parent, true);
     if (curStaticValue) {
       if ('value' in curStaticValue && typeof curStaticValue.value !== 'symbol' ||
-          typeof curStaticValue.then !== 'symbol' && typeof curStaticValue.else !== 'symbol') {
+          'then' in curStaticValue && typeof curStaticValue.then !== 'symbol' && typeof curStaticValue.else !== 'symbol') {
         staticChildValue = curStaticValue;
         staticChildNode = parent;
-        if (self.skip) self.skip();
+        if (self && self.skip) self.skip();
```

Comment:
I found that `self` was sometimes undefined. Is that expected?

Reply:
The WalkerContext should always be defined. Can you explain how it is ever coming up as undefined?

Reply:
I have no idea how I broke this. I must have changed `this` somehow but I can't seem to figure out where 🤔 

You can see the `asset-conditional` test fail without the check for `self` existence:

https://github.com/vercel/node-file-trace/pull/133/checks?check_run_id=862034673#step:8:111

Any ideas?

Reply:
Could this be a bug in the old version of `estree-walker` we are using?

I noticed that `enter` has the correct context but not `leave` which is where it is failing.

When I looked at the source code of `estree-walker` I see `enter.call(context, node, parent, prop, index);` and then `leave(node, parent, prop, index);` which makes me think its a bug in `estree-walker`.

Its still not clear to me why we didn't see this test failing before though 🤔

Reply:
I compared the stack trace: the left side is `master` and the right side is my PR.

![diff](https://user-images.githubusercontent.com/229881/87238815-9f858700-c3d5-11ea-9785-42c0b05dc136.png)

These are both using the same version of `estree-walker` 🤔

Reply:
Right, as you say the `leave` function does not have a skip option so that seems to be what is happening. So perhaps previously the `this` value was the global default and now it isn't or something like that. So, having the explicit context check for the enter case only makes sense as you have provided here.

Note that we can update the check though to `if (self)` instead of `if (self.skip)` since `self.skip` will be defined whenever `self` is defined, so we're just updating the entire check for the new `this` scoping.

Reply:
Thanks, I updated to `backtrack()` to accept the optional context and use `if (context) context.skip()`